### ISSUE TITLE
`return sorted(x)` does not infer type of `x`

### ISSUE THREAD

**Issue by jomido** [microsoft/pyright#2165]
**Describe the bug**
`sorted(x)` surfaces a type error when returning from a function.

**To Reproduce**
Using:
- VS Code version `1.59.0`
- pylance version `v2021.8.0`

Paste the code shown below. 

**Expected behavior**
Should not be a type error.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from dataclasses import dataclass
from typing import List


@dataclass
class Thing:
    pass


def hrm(things: List[Thing]) -> List[Thing]:

    return sorted(things)  # type error: No overloads for "sorted" match the provided arguments
```

**VS Code extension or command-line**
- VS Code version `1.59.0`
- pylance version `v2021.8.0`

**Additional context**
Related to #313.

**Comment by erictraut** [comment]
I think pyright is doing the right thing here given the type information it is provided. Mypy (another Python type checker) generates the same error.

The typeshed stub file `builtins.pyi` defines `sorted` as follows:

```python
@overload
def sorted(__iterable: Iterable[SupportsLessThanT], *, key: None = ..., reverse: bool = ...) -> List[SupportsLessThanT]: ...
@overload
def sorted(__iterable: Iterable[_T], *, key: Callable[[_T], SupportsLessThan], reverse: bool = ...) -> List[_T]: ...
```

You're not providing a value for `key`, so the first overload is the applicable one here. But the `Thing` class doesn't provide a `__lt__` method, so it does not meet the requirements of `SupportsLessThanT`. By default, dataclasses do not define an `__lt__` operator. If you add an `order=True` argument to the `@dataclass` decorator, a `__lt__` method will be synthesized for you, and it will type check without any errors.

```python
@dataclass(order=True)
class Thing:
    pass
```

**Comment by jomido** [comment]
Excellent explanation, thank you.

Interestingly enough, I derived a narrowed down code example from a real-world code snippet. That real-world code snippet _does_ define an `__lt__` method, but I removed it for the example. I'll put it back into the example:

```python
from dataclasses import dataclass
from typing import List


@dataclass
class Thing:
    def __lt__(self, other):

        return id(self) < id(other)


def hrm(things: List[Thing]) -> List[Thing]:

    return sorted(things)  # no type error
```

No type error! But my real-world code retains the type error, even though the `dataclass` `class` there defines an `__lt__`.

Digging a bit, I see the problem. Here is the example code again:

```python
from dataclasses import dataclass
from typing import List


@dataclass
class Thing:
    def __lt__(self, other):

        if id(self) < id(other):
            return True
        elif id(self) > id(other):
            return False


def hrm(things: List[Thing]) -> List[Thing]:

    return sorted(things)  # type error!
```

The `__lt__` method here contains a "bug": it doesn't check the `==` case.

The type inference/checking here is pretty damn amazing. Good job, Eric!

**Comment by jakebailey** [comment]
FWIW I think that if you had annotated that `__lt__` with `-> bool`, you should have gotten an error about the missing condition. Your last example relies on the implicit return of `None` being falsy, which may work for some expressions but surprise others that expect `bool`.

**Comment by jomido** [comment]
> FWIW I think that if you had annotated that __lt__ with -> bool, you should have gotten an error about the missing condition.

Thank you. I do get an error if I annotate the return type to `bool`. 👍 

> Your last example relies on the implicit return of None being falsy, which may work for some expressions but surprise others that expect bool.

Very nice - thank you.
